### PR TITLE
Added Auth0 device authorization flow example

### PR DESCRIPTION
**Pull request by schoenenberg**
Hi,

thank you for this great library! I have used it to implement a CLI using [Auth0](https://auth0.com). I would like to contribute this example back, because it took me some time to figure out a strange compiler error concerning the `request_async` method  for the `oauth2::DeviceAccessTokenRequest`.

Best regards,
Max

### CODE REVIEW

**Comment by Morishiri** [comment]
OMG. thank you very much for this!

Turns out the missing point to the one which already exists in the docs: https://docs.rs/oauth2/latest/oauth2/#device-code-flow

is 
```
    .set_auth_type(AuthType::RequestBody);
```

**Comment by michaeljones** [comment]
Thank you for the this example. It is very helpful.

File: examples/auth0_async_devicecode.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
+//!
+//! This example showcases the Auth0 device authorization flow using the async methods.
+//!
+//! Before running it, you'll need to create an API and an Application on [auth0.com](https://auth0.com). Take a look at this tutorial, for the details: [Call Your API Using the Device Authorization Flow](https://auth0.com/docs/flows/call-your-api-using-the-device-authorization-flow).
```
Comment (aminya) [inline examples/auth0_async_devicecode.rs:4]: This second link is dead now. 
```suggestion
//! Before running it, you'll need to create an API and an Application on [auth0.com](https://auth0.com). Take a look at this tutorial, for the details: [Call Your API Using the Device Authorization Flow](https://auth0.com/docs/get-started/authentication-and-authorization-flow/device-authorization-flow/call-your-api-using-the-device-authorization-flow).
```

**Review by aminya — CHANGES_REQUESTED** [review]

**Comment by schoenenberg** [comment]
Thanks @aminya! I don't know if this will ever be merged. But as long as it is helpful also in the PRs I will leave it here.

**Comment by ramosbugs** [comment]
> @ramosbugs I find this example very useful. Would you please consider merging it?

I'm trying to weigh the value of individual examples against the maintenance cost associated with them. I'm currently introducing some breaking changes to the next major release (e.g., #135, #236, #253), and updating all 9 current examples has been a pain.

The intent of this crate's examples is to illustrate how to use different features of this crate. Providing copy-pasteable working production code for every auth vendor is not a goal (though I would consider making exceptions for sponsors who help fund the maintenance of this project). That's the purview of vendor documentation/sample code, blog posts, etc. I'm willing to link to separately-maintained [contributed examples](https://github.com/ramosbugs/oauth2-rs/blob/1d1f4d17ecdf2a3feb565eb1789cc8649cac7705/src/lib.rs#L412-L414), which don't introduce maintenance load to this crate (other than link rot).

The specific features used by this example are:
- `exchange_device_code`/`exchange_device_access_token` (already illustrated by [`google_devicecode`](https://github.com/ramosbugs/oauth2-rs/blob/main/examples/google_devicecode.rs), [`microsoft_devicecode_common_user`](https://github.com/ramosbugs/oauth2-rs/blob/main/examples/microsoft_devicecode_common_user.rs), and [`microsoft_devicecode_tenant_user`](https://github.com/ramosbugs/oauth2-rs/blob/main/examples/microsoft_devicecode_tenant_user.rs))
- `set_auth_type(AuthType::RequestBody)` (already illustrated by [`google_devicecode`](https://github.com/ramosbugs/oauth2-rs/blob/main/examples/google_devicecode.rs), [`letterboxd`](https://github.com/ramosbugs/oauth2-rs/blob/main/examples/letterboxd.rs), and [`msgraph`](https://github.com/ramosbugs/oauth2-rs/blob/main/examples/msgraph.rs))
- `add_scope` (illustrated by 7 examples)
- `add_extra_param` (illusted by [`wunderlist`](https://github.com/ramosbugs/oauth2-rs/blob/main/examples/wunderlist.rs))
- device code flow with `request_async` (illustrated by [`microsoft_devicecode_common_user`](https://github.com/ramosbugs/oauth2-rs/blob/main/examples/microsoft_devicecode_common_user.rs) and [`microsoft_devicecode_tenant_user`](https://github.com/ramosbugs/oauth2-rs/blob/main/examples/microsoft_devicecode_tenant_user.rs))

Given that there are no new features illustrated by this example, I don't think it meets the criteria for adding a new example. It's reasonable to expect paying users of Auth0 who want to use this crate to read the documentation provided by Auth0 and the documentation provided by this crate and figure out how to use the two together. If an Auth0 user wants to publish a blog post or contribute an external example to help other Auth0 users, that would be great. However, it's unreasonable to expect unfunded libraries that implement standardized protocols to become comprehensive resources for interacting with every other entity that implements the same protocol.

**Comment by aminya** [comment]
That's fair. I was thinking of creating micro packages to handle common workflows like this. We can have crate like `oauth2-auth0`, and have the supported flows in that. Actually, all these URLs that are separate environment variables can be calculated using the Auth0 domain. So they can become even simpler.

**Comment by ramosbugs** [comment]
Vendor-specific packages built on top of this one (or `openidconnect`) sound good to me!

**Pull request closed without merging**